R/visit.R

```diff
@@ -43,3 +43,166 @@ visit_one <- function(pd_flat, funs) {
+#' @inherit context_towards_terminals
+#' @seealso context_towards_terminals visitors
+context_to_terminals <- function(pd_nested,
+                                 passed_lag_newlines,
+                                 passed_indent,
+                                 passed_spaces) {
+
+  if (is.null(pd_nested)) return()
+
+  pd_transformed <- context_towards_terminals(
+    pd_nested, passed_lag_newlines, passed_indent, passed_spaces
+  )
+
+  pd_transformed$child <- pmap(list(pd_transformed$child,
+                                    pd_transformed$lag_newlines,
+                                    pd_transformed$indent,
+                                    pd_transformed$spaces),
+                               context_to_terminals)
+  pd_transformed
+}
+
+
+#' Update the a parse table given outer context
+#'
+#' `passed_lag_newlines` are added to the first token in `pd`,
+#'   `passed_indent` is added to all tokens in `pd`, `passed_spaces` is added to
+#'   the last token in `pd`. [context_to_terminals()] calls this function
+#'   repeatedly, which means the propagation of the parse information to the
+#'   terminal tokens.
+#' @param pd_nested A nested parse table.
+#' @param passed_lag_newlines The lag_newlines to be propagated inwards.
+#' @param passed_indent The indention depth to be propagated inwards.
+#' @param passed_spaces The number of spaces to be propagated inwards.
+#' @return An updated parse table.
+#' @seealso context_to_terminals
+context_towards_terminals <- function(pd_nested,
+                                      passed_lag_newlines,
+                                      passed_indent,
+                                      passed_spaces) {
+  pd_nested$indent <- pd_nested$indent + passed_indent
+  pd_nested$lag_newlines[1] <- pd_nested$lag_newlines[1] + passed_lag_newlines
+  pd_nested$spaces[nrow(pd_nested)] <-
+    pd_nested$spaces[nrow(pd_nested)] + passed_spaces
+  pd_nested
+}
+
+#' Extract terminal tokens
+#'
+#' Turns a nested parse table into a flat parse table. In particular it extracts
+#'   terminal tokens and the following attributes:
+#'
+#'  * lag_newlines
+#'  * indent
+#'  * token
+#'  * text
+#'  * spaces
+#'  * id
+#'  * parent
+#'  * line1
+#' @inheritParams extract_terminals_helper
+#' @importFrom readr type_convert col_integer cols
+extract_terminals <- function(pd_nested) {
+  flat_vec <- extract_terminals_helper(pd_nested) %>%
+    unlist()
+  nms <- list(
+    NULL,
+    c("lag_newlines", "indent", "token", "text", "spaces", "id", "parent", "line1")
+  )
+  flat_tbl <- matrix(flat_vec, ncol = length(nms[[2]]), byrow = TRUE, dimnames = nms) %>%
+    as_tibble() %>%
+    type_convert(
+      col_types = cols(
+        lag_newlines = col_integer(),
+        indent       = col_integer(),
+        spaces       = col_integer()
+      )
+    )
+}
+
+#' Helper to extract terminals
+#'
+#' @param pd_nested A nested parse table.
+extract_terminals_helper <- function(pd_nested) {
```

Comment:
Instead of sorting by (line1, col1), we could create our own sequential sort key; new tokens could then get intermediate values (e.g., for the closing parens, key + 1/3 and key + 2/3).